torch/utils/data/_typing.py

```diff
@@ -338,6 +344,59 @@ def __hash__(self):
         return hash((self.__name__, self.type))
 
 
+def hook_iterator(namespace, profile_name):
+
+    def context():
+        return torch.autograd.profiler.record_function(profile_name)
+
+    class IteratorDecorator:
+        '''Wrap the iterator return result by adding __next__'''
+        def __init__(self, iterator):
+            self.iterator = iterator
+
+        def __iter__(self):
+            return self
+
+        def __next__(self):
+            with context():
+                return next(self.iterator)
+
+    func = namespace['__iter__']
+
+    if inspect.isgeneratorfunction(func):
+        @functools.wraps(func)
+        def wrap_generator(*args, **kwargs):
+            gen = func(*args, **kwargs)
+            try:
+                with context():
+                    response = gen.send(None)
+                while True:
+                    request = yield response
+                    with context():
+                        response = gen.send(request)
+            except StopIteration as e:
```

Comment:
Another thing I am thinking why not:
```py
def wrap_generator(*args, **kwargs):
    return IteratorDecorator(func(*args, **kwargs))
```

Reply:
the reason that I don't use IteratorDecorator for generator: we'd like the behavior of __iter__ after decorated is still generator. Otherwise, the generator behavior will be changed as function which might break the compile test cases.